google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/ReadWriteTransaction.java

```diff
@@ -0,0 +1,1283 @@
+  private final List<RetriableStatement> statements = new ArrayList<>();
+  private final List<Mutation> mutations = new ArrayList<>();
+  private Timestamp transactionStarted;
+
+  static class Builder extends AbstractMultiUseTransaction.Builder<Builder, ReadWriteTransaction> {
+    private DatabaseClient dbClient;
+    private Boolean retryAbortsInternally;
+    private List<TransactionRetryListener> transactionRetryListeners;
+
+    private Builder() {}
+
+    Builder setDatabaseClient(DatabaseClient client) {
+      Preconditions.checkNotNull(client);
+      this.dbClient = client;
+      return this;
+    }
+
+    Builder setRetryAbortsInternally(boolean retryAbortsInternally) {
+      this.retryAbortsInternally = retryAbortsInternally;
+      return this;
+    }
+
+    Builder setTransactionRetryListeners(List<TransactionRetryListener> listeners) {
+      Preconditions.checkNotNull(listeners);
+      this.transactionRetryListeners = listeners;
+      return this;
+    }
+
+    @Override
+    ReadWriteTransaction build() {
+      Preconditions.checkState(dbClient != null, "No DatabaseClient client specified");
+      Preconditions.checkState(
+          retryAbortsInternally != null, "RetryAbortsInternally is not specified");
+      Preconditions.checkState(
+          transactionRetryListeners != null, "TransactionRetryListeners are not specified");
+      return new ReadWriteTransaction(this);
+    }
+  }
+
+  static Builder newBuilder() {
+    return new Builder();
+  }
+
+  private ReadWriteTransaction(Builder builder) {
+    super(builder);
+    this.transactionId = ID_GENERATOR.incrementAndGet();
+    this.dbClient = builder.dbClient;
+    this.retryAbortsInternally = builder.retryAbortsInternally;
+    this.transactionRetryListeners = builder.transactionRetryListeners;
+    this.txManager = dbClient.transactionManager();
+  }
+
+  @Override
+  public String toString() {
+    return new StringBuilder()
+        .append("ReadWriteTransaction - ID: ")
+        .append(transactionId)
+        .append("; Status: ")
+        .append(internalGetStateName())
+        .append("; Started: ")
+        .append(internalGetTimeStarted())
+        .append("; Retry attempts: ")
+        .append(transactionRetryAttempts)
+        .append("; Successful retries: ")
+        .append(successfulRetries)
+        .toString();
+  }
+
+  private String internalGetStateName() {
+    return transactionStarted == null ? "Not yet started" : getState().toString();
+  }
+
+  private String internalGetTimeStarted() {
+    return transactionStarted == null ? "Not yet started" : transactionStarted.toString();
+  }
+
+  @Override
+  public UnitOfWorkState getState() {
+    return this.state;
+  }
+
+  @Override
+  public boolean isReadOnly() {
+    return false;
+  }
+
+  @Override
+  void checkValidTransaction() {
+    ConnectionPreconditions.checkState(
+        state == UnitOfWorkState.STARTED, "This transaction has status " + state.name());
```

Comment:
Would slightly change the precondition message to add acceptable status

Reply:
done